### ISSUE TITLE
Lint for empty statements (was Linter should catch `if (foo);`)

### ISSUE THREAD

**Issue by Hixie** [dart-lang/linter#259]
I just got caught by the following code:

```dart
  if (complicated.expression.foo());
    bar();
```

It would be great if the linter could catch cases of an "if" block doing nothing.

**Comment by alexeieleusis** [comment]
Some code styles prevent this by requiring `if` statements to always have bodies. Don't you like that approach?

**Comment by Hixie** [comment]
How would that stop this case?

```dart
if (complicated.expression.foo()); {
    bar();
}
```

...has the same issue.

**Comment by pq** [comment]
If I'm not missing something I think both of those cases are essentially the same at the AST level: the rub is that the `if` has an empty `then` statement.

Incidentally, using a formatter really helps make these cases more obvious as `dartftmt` wrestles your examples into these respectively.

```dart
  if (complicated.expression.foo()) ;
  bar();
```

```dart
if (complicated.expression.foo()) ; 
{
    bar();
}
```


But anyway, I think a rule checking for empty then statement is totally reasonable.

Curious if @bwilkerson cares to chime in?  It's early.  First cup of coffee, so I'm sure I'm overlooking something!

☕

**Comment by pq** [comment]
FWIW: I'm playing with an implementation.  (We can always toss it if we come to another conclusion on it's value.)

**Comment by bwilkerson** [comment]
Personally, I think it's always questionable to have an empty statement, so I would just flag empty statements where ever they occur. I'm kind of surprised that we don't already have that somewhere.

**Comment by pq** [comment]
I'm with Brian.

**Comment by Hixie** [comment]
Totally agree that we should just catch empty statements.

(And yeah, a formatter would help here too. But that assumes you look at the formatted code...)
